chalk-solve/src/clauses.rs

```diff
@@ -277,6 +293,36 @@ fn program_clauses_that_could_match<I: Interner>(
             let associated_ty_datum = db.associated_ty_data(alias.associated_ty_id);
             let trait_id = associated_ty_datum.trait_id;
             let trait_parameters = db.trait_parameters_from_projection(alias);
+
+            if (alias
+                .self_type_parameter(interner)
+                .bound(interner)
```

Comment:
it seems like passing in bound variables is causing a lot of code duplication. At minimum, we should make a helper, but I'm semi-inclined not to pass them in in the first place. But I'll have to see when I get to the code below. =)

Reply:
OK, so, I understand why this is setup this way now. I'm a bit unsure about it. For one thing, I'm not sure whether bound variables in canonicalized goals should *exclusively* represent inference variables, though that's the case right now. In rustc, canonicalization also "renumbers and replaces" placeholders, and the more I think about it, the more I think that makes sense. 

I guess that to change this would require creating the inference context earlier and leveraging checkpoints. i.e., we would create the inference table in `solve_from_clauses` and then clone it for each `Fulfill::new` call (or else, if not cloning, using a snapshot).

Reply:
But I think we could address this in a follow-up issue.

Reply:
added note to https://github.com/rust-lang/chalk/issues/351